lib/3scale_toolbox/commands/product_command/copy_command.rb

```diff
@@ -0,0 +1,78 @@
+require '3scale_toolbox/commands/product_command/copy_command/delete_target_backend_usages_task'
+require '3scale_toolbox/commands/product_command/copy_command/copy_backends_task'
+
+module ThreeScaleToolbox
+  module Commands
+    module ProductCommand
+      class CopySubcommand < Cri::CommandRunner
+        include ThreeScaleToolbox::Command
+
+        def self.command
+          Cri::Command.define do
+            name        'copy'
+            usage       'copy [opts] -s <source_remote> -d <target_remote> <source_product>'
+            summary     'Copy product'
+            description <<-HEREDOC
+            This command makes a copy of the referenced product.
+            Target product will be searched by source product system name. System name can be overriden with `--target_system_name` option.
+            If a product with the selected `system_name` is not found, it will be created.
+            \n Components of the product being copied:
+            \nproduct configuration
+            \nproduct settings
+            \nproduct methods&metrics
+            \nproduct mapping rules
+            \nproduct application plans & pricing rules & limits
+            \nproduct application usage rules
+            \nproduct policies
+            \nproduct backends
+            \nproduct activedocs
+            HEREDOC
+
+            option  :s, :source, '3scale source instance. Url or remote name', argument: :required
+            option  :d, :destination, '3scale target instance. Url or remote name', argument: :required
+            option  :t, 'target_system_name', 'Target system name. Default to source system name', argument: :required
+            param   :source_product
+
+            runner CopySubcommand
+          end
+        end
+
+        def run
+          tasks = []
+          tasks << ThreeScaleToolbox::Commands::ServiceCommand::CopyCommand::CreateOrUpdateTargetServiceTask.new(context)
+          tasks << CopyCommand::DeleteTargetBackendUsagesTask.new(context)
```

Comment:
Why do we delete the target backend usages before copying the source ones?

Reply:
Because 3scal API only allows updating path of a backend usage. So, if the path is already in use, there is no way to update target backend (backend_api_id attr of the backend usage object).

Hence, delete first approach has been followed. Happy to hear any other alternative.

Reply:
I made the original question due to I think the following scenario might happen with the current appraoch and I'm not sure that's what we want: If the destination product already exists, it might have backend ussages associated to backends that only exist on the destination but not in the source, and those would also be deleted during the copy and therefore lost. What do you think?

An alternative would be:

* Leave untouched the existing backendusages on the destination that don't have a matching backend system name (taken into account that this copy command itself might introduce them) and delete and recreate the backendusages that have a match.

Reply:
Existing backend usages in the destination product that do not conflict (by path) with origin product backend usages are left and not deleted.

Your approach would work but it would require to build mapping between source backends and target backends. To save building this mapping, the implemented approach is:

> Existing backend usages in the destination product that do not conflict (by path) with origin product backend usages are left and not deleted.

This way, all source backend usages can be created safely in target product. Besides, unrelated target backend usages are not deleted.

Reply:
:+1: